fixed forced destroy of securitylake when meta_store_role_arn changes

joelmccoy opened the pull request:
<!---
See what makes a good Pull Request at: https://hashicorp.github.io/terraform-provider-aws/raising-a-pull-request/
--->
### Description
<!---
Please provide a helpful description of what change this pull request will introduce.
--->
This change updates the aws_securitylake_data_lake logic so that when the attribute `meta_store_manager_role_arn` is changed it does not require a delete of the entire data lake.  The [UpdateDataLake](https://docs.aws.amazon.com/security-lake/latest/APIReference/API_UpdateDataLake.html) recently included a change where you can update the metaStoreManagerRoleArn with this API call.  The stringplanmodifier.RequiresReplace is no longer necessary.

### Relations
<!---
If your pull request fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates.

For Example:

Relates #0000
or 
Closes #0000
--->

Closes #36831 

### References
<!---
Optionally, provide any helpful references that may help the reviewer(s).
--->
N/A

### Output from Acceptance Testing
<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

Replace ec2 with the service package corresponding to your tests.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->

```console
$  make testacc TESTS=TestAccSecurityLake_ PKG=securitylake
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go1.21.8 test ./internal/service/securitylake/... -v -count 1 -parallel 20 -run='TestAccSecurityLake_'  -timeout 360m
=== RUN   TestAccSecurityLake_serial
=== PAUSE TestAccSecurityLake_serial
=== CONT  TestAccSecurityLake_serial
=== RUN   TestAccSecurityLake_serial/DataLake
=== RUN   TestAccSecurityLake_serial/DataLake/metaStoreUpdate
=== RUN   TestAccSecurityLake_serial/DataLake/replication
=== RUN   TestAccSecurityLake_serial/DataLake/basic
=== RUN   TestAccSecurityLake_serial/DataLake/disappears
=== RUN   TestAccSecurityLake_serial/DataLake/tags
=== RUN   TestAccSecurityLake_serial/DataLake/lifecycle
=== RUN   TestAccSecurityLake_serial/DataLake/lifecycleUpdate
=== RUN   TestAccSecurityLake_serial/Subscriber
=== RUN   TestAccSecurityLake_serial/Subscriber/basic
=== RUN   TestAccSecurityLake_serial/Subscriber/customLogs
=== RUN   TestAccSecurityLake_serial/Subscriber/disappears
=== RUN   TestAccSecurityLake_serial/Subscriber/tags
=== RUN   TestAccSecurityLake_serial/Subscriber/updated
=== RUN   TestAccSecurityLake_serial/SubscriberNotification
=== RUN   TestAccSecurityLake_serial/SubscriberNotification/https
=== RUN   TestAccSecurityLake_serial/SubscriberNotification/disappears
=== RUN   TestAccSecurityLake_serial/SubscriberNotification/update
=== RUN   TestAccSecurityLake_serial/SubscriberNotification/basic
=== RUN   TestAccSecurityLake_serial/AWSLogSource
=== RUN   TestAccSecurityLake_serial/AWSLogSource/basic
=== RUN   TestAccSecurityLake_serial/AWSLogSource/disappears
=== RUN   TestAccSecurityLake_serial/AWSLogSource/multiRegion
=== RUN   TestAccSecurityLake_serial/CustomLogSource
=== RUN   TestAccSecurityLake_serial/CustomLogSource/basic
=== RUN   TestAccSecurityLake_serial/CustomLogSource/disappears
--- PASS: TestAccSecurityLake_serial (1141.58s)
    --- PASS: TestAccSecurityLake_serial/DataLake (365.31s)
        --- PASS: TestAccSecurityLake_serial/DataLake/metaStoreUpdate (55.85s)
        --- PASS: TestAccSecurityLake_serial/DataLake/replication (80.38s)
        --- PASS: TestAccSecurityLake_serial/DataLake/basic (35.53s)
        --- PASS: TestAccSecurityLake_serial/DataLake/disappears (34.71s)
        --- PASS: TestAccSecurityLake_serial/DataLake/tags (57.93s)
        --- PASS: TestAccSecurityLake_serial/DataLake/lifecycle (40.68s)
        --- PASS: TestAccSecurityLake_serial/DataLake/lifecycleUpdate (60.21s)
    --- PASS: TestAccSecurityLake_serial/Subscriber (265.16s)
        --- PASS: TestAccSecurityLake_serial/Subscriber/basic (43.56s)
        --- PASS: TestAccSecurityLake_serial/Subscriber/customLogs (44.93s)
        --- PASS: TestAccSecurityLake_serial/Subscriber/disappears (40.37s)
        --- PASS: TestAccSecurityLake_serial/Subscriber/tags (76.45s)
        --- PASS: TestAccSecurityLake_serial/Subscriber/updated (59.85s)
    --- PASS: TestAccSecurityLake_serial/SubscriberNotification (275.65s)
        --- PASS: TestAccSecurityLake_serial/SubscriberNotification/https (55.18s)
        --- PASS: TestAccSecurityLake_serial/SubscriberNotification/disappears (56.28s)
        --- PASS: TestAccSecurityLake_serial/SubscriberNotification/update (109.87s)
        --- PASS: TestAccSecurityLake_serial/SubscriberNotification/basic (54.31s)
    --- PASS: TestAccSecurityLake_serial/AWSLogSource (155.59s)
        --- PASS: TestAccSecurityLake_serial/AWSLogSource/basic (41.52s)
        --- PASS: TestAccSecurityLake_serial/AWSLogSource/disappears (39.81s)
        --- PASS: TestAccSecurityLake_serial/AWSLogSource/multiRegion (74.27s)
    --- PASS: TestAccSecurityLake_serial/CustomLogSource (79.87s)
        --- PASS: TestAccSecurityLake_serial/CustomLogSource/basic (39.63s)
        --- PASS: TestAccSecurityLake_serial/CustomLogSource/disappears (40.24s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/securitylake       1146.023s

...
```

gdavison commented on internal/service/securitylake/data_lake_test.go:
```diff
@@ -344,6 +390,31 @@ const testAccDataLakeConfigConfig_base = `
 data "aws_caller_identity" "current" {}
 data "aws_partition" "current" {}
 
+resource "aws_iam_role" "meta_store_manager_v1" {
+  name               = "AmazonSecurityLakeMetaStoreManagerV1"
+  path               = "/service-role/"
+  assume_role_policy = <<POLICY
+{
+  "Version": "2012-10-17",
+  "Statement": [{
+    "Sid": "AllowLambda",
+    "Effect": "Allow",
+    "Principal": {
+      "Service": [
+        "lambda.amazonaws.com"
+      ]
+    },
+    "Action": "sts:AssumeRole"
+  }]
+}
+POLICY
+}
+
+resource "aws_iam_role_policy_attachment" "datalake_v1" {
+  role       = aws_iam_role.meta_store_manager_v1.name
+  policy_arn = "arn:${data.aws_partition.current.partition}:iam::aws:policy/service-role/AmazonSecurityLakeMetastoreManager"
+}
+
```
Since this Role is only used in the `testAccDataLakeConfig_metaStore` and `testAccDataLakeConfig_metaStoreUpdate`, it would be better to only include it in those configurations, rather than for all tests

gdavison commented on internal/service/securitylake/data_lake_test.go:
```diff
@@ -581,6 +652,43 @@ resource "aws_securitylake_data_lake" "test" {
+    Name = %[1]q
+  }
+
+  depends_on = [aws_iam_role.meta_store_manager]
```
Since `meta_store_manager_role_arn` already refers to `aws_iam_role.meta_store_manager`, this `depends_on` is redundant.

gdavison commented on internal/service/securitylake/data_lake_test.go:
```diff
@@ -581,6 +652,43 @@ resource "aws_securitylake_data_lake" "test" {
+    region = %[2]q
+  }
+
+  tags = {
+    Name = %[1]q
+  }
```
`tags` can be removed, since it's not relevant to this test

gdavison reviewed (changes requested):
Thanks for the PR, @joelmccoy. It looks good. I have a few suggestions

joelmccoy commented:
@gdavison Good feedback.  I implemented the changes... but in the process I discovered an issue with the [UpdateDataLake](https://docs.aws.amazon.com/security-lake/latest/APIReference/API_UpdateDataLake.html) api call.  When you specify a new role it just adds this new role to the Data Lake Administrators list (in the LakeFormation Service) but it doesn't remove the previous role.  

Essentially the SecurityLake API is doing some LakeFormation calls under the hood and it is leaving dangling resources in LakeFormation that can potentially cause some issues.

Here is an example of the flow:

1. CreateDataLake with role1 (role1 is admin)
2. UpdateDataLake with role2 (role1 and role2 are both admin)

This becomes an issue because if you delete the IAM role1 or role2 ALL SecurityLake operations will fail (yikes). I discovered this in my tests when i removed the v1 role out of the basic config.  It seems to check this Lakeformation admin list and will fail if any of the admin IAM roles do not exist.  I think this is also the case for Deleting Security Lake Resource.  It seems to leave this dangling Admin Assignment in Lakeformation. 

The only way to change these admins I believe is through this LakeFormation [PutDataLakeSettings](https://docs.aws.amazon.com/lake-formation/latest/APIReference/API_PutDataLakeSettings.html).

A hacky way around this is would be to make some lakeformation calls in the terraform provider to do the cleanup on a meta_store_role_arn change.  But that doesn't feel right and I think this should be fixed upstream in the APIs.  Agree/disagree?

gdavison commented:
Good find, @joelmccoy. We do have a number of cases where we need to do some cleanup to "help" AWS services. For example, `deleteLingeringENIs` at https://github.com/hashicorp/terraform-provider-aws/blob/aaedd7d9bf0c920b2e7198fef4578142f21a1f5c/internal/service/ec2/vpc_network_interface.go#L1453 needs to clean up network interfaces that sometimes take a long time to be cleaned up automatically.

Do you know if there's a delayed process on the AWS side that eventually removes the role from Lake Formation? If so, we could wait for that. Otherwise, the provider may have to reach across service boundaries.

joelmccoy commented:
@gdavison It doesn't seem to ever get deleted.  If it is a sane approach to add a helper to clean up lingering
Lakeformation admin role assignments I can definitely take a stab at adding that logic.
